# <center>💟INFORME MACHING LEARNING💟</center>

El objetivo de este informe es crear dos notebooks. Uno donde se utilice la técnica de redes neuronales
y otro para la técnica de árboles de decisión, todo esto para determinar los hiperparámetros óptimos para obtener la mayor precisión posible.

💟INTEGRANTES💟
* Isabela Rosero 2128720
* Luisa Cardenas 1823494 
* Stefhania Noguera 2125854

💟IMPORTACIONES💟

In [52]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import HistGradientBoostingClassifier

💟LEER EL ARCHIVO CSV💟

In [53]:
# 1: Leer el archivo breastCancer.csv

data = pd.read_csv("breastCancer.csv")

💟SELECCIONAR EL 80% DEL CONJUNTO DE DATOS Y EL 20 PARA PRUEBAS💟

In [54]:
# 2: Seleccionar aleatoriamente el 80% del conjunto de datos para entrenar y el 20% restante para las pruebas
X = data.drop("diagnosis", axis=1)
y = data["diagnosis"]
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=123
) 

💟NORMALIZAR LOS DATOS💟

In [55]:
numericas = ["radius_mean",
              "texture_mean", 
              "perimeter_mean", 
              "area_mean", 
              "smoothness_mean", 
              "compactness_mean", 
              "symmetry_mean", 
              "fractal_dimension_mean", 
              "radius_worst",
              "texture_worst",
              "perimeter_worst",
              "area_worst",
              "smoothness_worst",
              "compactness_worst",
              "symmetry_worst",
              "fractal_dimension_worst",]


num_transformer = Pipeline(
    steps=[("imputer", SimpleImputer(strategy="median")), ("scaler", StandardScaler())]
)

modificacionesPreProcesado = ColumnTransformer(
    transformers=[
        ("num", num_transformer, numericas),
    ]
)

X_train_processed = modificacionesPreProcesado.fit_transform(X_train)
X_test_processed = modificacionesPreProcesado.transform(X_test)

💟CONFIGURAR LOS HPERPARAMETROS DEL ARBOL💟

In [56]:
hiperparametro = range(1, 11, 1)

arbolesDecisionPaso4 = []
for max_depth in hiperparametro:
    classifier = DecisionTreeClassifier(
        criterion="gini", splitter="best", max_depth=max_depth, random_state=123
    )
    classifier.fit(X_train_processed, y_train)
    arbolesDecisionPaso4.append(classifier)

💟TABLA CON LOS VALORES💟

In [57]:
valoresPrecisionPaso4 = []
for max_depth, tree in zip(hiperparametro, arbolesDecisionPaso4):
    precision = tree.score(X_test_processed, y_test)
    valoresPrecisionPaso4.append(precision)

tablaPaso4 = pd.DataFrame(
    {"Profundidad máxima": hiperparametro, "Precisión": valoresPrecisionPaso4}
)
print("Tabla de precisión de árboles punto 4")
print(tablaPaso4, "\n")

Tabla de precisión de árboles punto 4
   Profundidad máxima  Precisión
0                   1   0.929825
1                   2   0.929825
2                   3   0.956140
3                   4   0.947368
4                   5   0.964912
5                   6   0.964912
6                   7   0.956140
7                   8   0.964912
8                   9   0.964912
9                  10   0.964912 



💟CAMBIOS EN LOS HYPERPARAMETROS💟

In [58]:
arbolesDecisionPaso6 = []
for max_depth in hiperparametro:
    classifier = DecisionTreeClassifier(
        criterion="entropy", splitter="best", max_depth=max_depth, random_state=123
    )
    classifier.fit(X_train_processed, y_train)
    arbolesDecisionPaso6.append(classifier)

💟TABLA CON SUS VALORES💟

In [59]:
valoresPrecisionPaso6 = []
for max_depth, tree in zip(hiperparametro, arbolesDecisionPaso6):
    accuracy = tree.score(X_test_processed, y_test)
    valoresPrecisionPaso6.append(accuracy)

tablaPaso6 = pd.DataFrame(
    {"Profundidad máxima": hiperparametro, "Precisión": valoresPrecisionPaso6}
)
print("Tabla de precisión de árboles punto 6")
print(tablaPaso6, "\n")

Tabla de precisión de árboles punto 6
   Profundidad máxima  Precisión
0                   1   0.929825
1                   2   0.929825
2                   3   0.973684
3                   4   0.964912
4                   5   0.964912
5                   6   0.964912
6                   7   0.991228
7                   8   0.991228
8                   9   0.991228
9                  10   0.991228 



💟3 CAMBIO EN LOS HYPERPARAMETROS💟

In [60]:
arbolesDecisionPaso8 = []
for max_depth in hiperparametro:
    classifier = HistGradientBoostingClassifier(max_depth=max_depth, random_state=123)
    classifier.fit(X_train_processed, y_train)
    arbolesDecisionPaso8.append(classifier)


💟TABLA CON SUS VALORES💟

In [61]:
valoresPrecisionPaso8 = []
for max_depth, tree in zip(hiperparametro, arbolesDecisionPaso8):
    accuracy = tree.score(X_test_processed, y_test)
    valoresPrecisionPaso8.append(accuracy)

tablaPaso8 = pd.DataFrame(
    {"Profundidad máxima": hiperparametro, "Precisión": valoresPrecisionPaso8}
)
print("Tabla de precisión de árboles punto 8")
print(tablaPaso8, "\n")

Tabla de precisión de árboles punto 8
   Profundidad máxima  Precisión
0                   1   0.973684
1                   2   0.982456
2                   3   0.973684
3                   4   0.982456
4                   5   0.982456
5                   6   0.982456
6                   7   0.982456
7                   8   0.982456
8                   9   0.982456
9                  10   0.982456 



💟ENCONTRAMOS EL MEJOR ARBOL💟

In [62]:

arboles = [
    max(arbolesDecisionPaso4, key=lambda tree: tree.score(X_test_processed, y_test)),
    max(arbolesDecisionPaso6, key=lambda tree: tree.score(X_test_processed, y_test)),
    max(arbolesDecisionPaso8, key=lambda tree: tree.score(X_test_processed, y_test)),
]


mejor_arbol = max(arboles, key=lambda tree: tree.score(X_test_processed, y_test))

if arboles.index(mejor_arbol) == 0:
    print("El mejor árbol es el del punto 4")
elif arboles.index(mejor_arbol) == 1:
    print("El mejor árbol es el del punto 6")
else:
    print("El mejor árbol es el del punto 8")

mejor_hiperparametros = mejor_arbol.get_params()
print("Hiperparámetros del mejor árbol:")
print(mejor_hiperparametros)
print("Mejor arbol accuracy:", mejor_arbol.score(X_test_processed, y_test))

El mejor árbol es el del punto 6
Hiperparámetros del mejor árbol:
{'ccp_alpha': 0.0, 'class_weight': None, 'criterion': 'entropy', 'max_depth': 7, 'max_features': None, 'max_leaf_nodes': None, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'random_state': 123, 'splitter': 'best'}
Mejor arbol accuracy: 0.9912280701754386


💟SE CAMBIAN LOS PARAMETROS PARA VERIFICAR💟

In [63]:
classifier_variacion1 = DecisionTreeClassifier(
    criterion=mejor_hiperparametros["criterion"],
    splitter=mejor_hiperparametros["splitter"],
    max_depth=mejor_hiperparametros["max_depth"],
    random_state=mejor_hiperparametros["random_state"],
    max_leaf_nodes=5,
)

classifier_variacion1.fit(X_train_processed, y_train)
accuracy_variacion1 = classifier_variacion1.score(X_test_processed, y_test)
print("Accuracy con max_leaf_nodes=2:", accuracy_variacion1)

# Variación 2: max_leaf_nodes = 10
classifier_variacion2 = DecisionTreeClassifier(
    criterion=mejor_hiperparametros["criterion"],
    splitter=mejor_hiperparametros["splitter"],
    max_depth=mejor_hiperparametros["max_depth"],
    random_state=mejor_hiperparametros["random_state"],
    max_leaf_nodes=10,
)

classifier_variacion2.fit(X_train_processed, y_train)
accuracy_variacion2 = classifier_variacion2.score(X_test_processed, y_test)
print("Accuracy con max_leaf_nodes=10:", accuracy_variacion2)

Accuracy con max_leaf_nodes=2: 0.9649122807017544
Accuracy con max_leaf_nodes=10: 0.9736842105263158


# <center>💟CODIGO COMPLETO💟</center>

In [51]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import HistGradientBoostingClassifier

# 1: Leer el archivo breastCancer.csv

data = pd.read_csv("breastCancer.csv")

# 2: Seleccionar aleatoriamente el 80% del conjunto de datos para entrenar y el 20% restante para las pruebas
X = data.drop("diagnosis", axis=1)
y = data["diagnosis"]
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=123
) 

# 3: Utilizar una estrategia para normalizar los datos.
# Codificación one-hot para variables categóricas y transformaciones para variables numéricas
numericas = ["radius_mean",
              "texture_mean", 
              "perimeter_mean", 
              "area_mean", 
              "smoothness_mean", 
              "compactness_mean", 
              "symmetry_mean", 
              "fractal_dimension_mean", 
              "radius_worst",
              "texture_worst",
              "perimeter_worst",
              "area_worst",
              "smoothness_worst",
              "compactness_worst",
              "symmetry_worst",
              "fractal_dimension_worst",]


num_transformer = Pipeline(
    steps=[("imputer", SimpleImputer(strategy="median")), ("scaler", StandardScaler())]
)

modificacionesPreProcesado = ColumnTransformer(
    transformers=[
        ("num", num_transformer, numericas),
    ]
)

X_train_processed = modificacionesPreProcesado.fit_transform(X_train)
X_test_processed = modificacionesPreProcesado.transform(X_test)

# 4: Configurar los hiperparámetros del árbol de decisión de la siguiente manera: criterion=’gini’,
#splitter=’best’, y random_state=123. Obtener 10 árboles de decisión que resultan de modificar el hiperparámetro max_depth desde 1 hasta 10 de 1 en 1.

hiperparametro = range(1, 11, 1)

arbolesDecisionPaso4 = []
for max_depth in hiperparametro:
    classifier = DecisionTreeClassifier(
        criterion="gini", splitter="best", max_depth=max_depth, random_state=123
    )
    classifier.fit(X_train_processed, y_train)
    arbolesDecisionPaso4.append(classifier)

# 5: Incluya en el notebook una tabla con el accuracy para los 10 árboles del punto anterior

valoresPrecisionPaso4 = []
for max_depth, tree in zip(hiperparametro, arbolesDecisionPaso4):
    precision = tree.score(X_test_processed, y_test)
    valoresPrecisionPaso4.append(precision)

tablaPaso4 = pd.DataFrame(
    {"Profundidad máxima": hiperparametro, "Precisión": valoresPrecisionPaso4}
)
print("Tabla de precisión de árboles punto 4")
print(tablaPaso4, "\n")

# 6: Repita el mismo procedimiento del punto 4 usando como hiperparámetros criterion=’entropy’,splitter=’best’, random_state=123,
# y variando el hiperparámetro max_depth desde 1 hasta 10 de 1 en 1.

arbolesDecisionPaso6 = []
for max_depth in hiperparametro:
    classifier = DecisionTreeClassifier(
        criterion="entropy", splitter="best", max_depth=max_depth, random_state=123
    )
    classifier.fit(X_train_processed, y_train)
    arbolesDecisionPaso6.append(classifier)

# 7: Incluya en el notebook una tabla con el accuracy para los 10 árboles del punto anterior.

valoresPrecisionPaso6 = []
for max_depth, tree in zip(hiperparametro, arbolesDecisionPaso6):
    accuracy = tree.score(X_test_processed, y_test)
    valoresPrecisionPaso6.append(accuracy)

tablaPaso6 = pd.DataFrame(
    {"Profundidad máxima": hiperparametro, "Precisión": valoresPrecisionPaso6}
)
print("Tabla de precisión de árboles punto 6")
print(tablaPaso6, "\n")

# 8: Repita el mismo procedimiento del punto 4 usando como hiperparámetros criterion=’entropy’, splitter=’random’, random_state=123,
# y variando el hiperparámetro max_depth desde 1 hasta 10 de 1 en 1.

arbolesDecisionPaso8 = []
for max_depth in hiperparametro:
    classifier = HistGradientBoostingClassifier(max_depth=max_depth, random_state=123)
    classifier.fit(X_train_processed, y_train)
    arbolesDecisionPaso8.append(classifier)

# 9: Incluya en el notebook una tabla con el accuracy para los 10 árboles del punto anterior.
    
valoresPrecisionPaso8 = []
for max_depth, tree in zip(hiperparametro, arbolesDecisionPaso8):
    accuracy = tree.score(X_test_processed, y_test)
    valoresPrecisionPaso8.append(accuracy)

tablaPaso8 = pd.DataFrame(
    {"Profundidad máxima": hiperparametro, "Precisión": valoresPrecisionPaso8}
)
print("Tabla de precisión de árboles punto 8")
print(tablaPaso8, "\n")

# 10: Indique en el notebook los hiperparámetros que por el momento le permiten obtener el árbol con mayor accuracy

arboles = [
    max(arbolesDecisionPaso4, key=lambda tree: tree.score(X_test_processed, y_test)),
    max(arbolesDecisionPaso6, key=lambda tree: tree.score(X_test_processed, y_test)),
    max(arbolesDecisionPaso8, key=lambda tree: tree.score(X_test_processed, y_test)),
]


mejor_arbol = max(arboles, key=lambda tree: tree.score(X_test_processed, y_test))

if arboles.index(mejor_arbol) == 0:
    print("El mejor árbol es el del punto 4")
elif arboles.index(mejor_arbol) == 1:
    print("El mejor árbol es el del punto 6")
else:
    print("El mejor árbol es el del punto 8")

mejor_hiperparametros = mejor_arbol.get_params()
print("Hiperparámetros del mejor árbol:")
print(mejor_hiperparametros)
print("Mejor arbol accuracy:", mejor_arbol.score(X_test_processed, y_test))

# 11: Seleccione uno de los hiperparámetros disponibles en la documentación (https://scikitlearn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html) que sea diferente al criterion,
#splitter, max_depth, y random_state. Realice dos variaciones en el hiperparámetro seleccionado manteniendo los otros hiperparámetros del punto anterior

# Variación 1: max_leaf_nodes = 5
classifier_variacion1 = DecisionTreeClassifier(
    criterion=mejor_hiperparametros["criterion"],
    splitter=mejor_hiperparametros["splitter"],
    max_depth=mejor_hiperparametros["max_depth"],
    random_state=mejor_hiperparametros["random_state"],
    max_leaf_nodes=5,
)

classifier_variacion1.fit(X_train_processed, y_train)
accuracy_variacion1 = classifier_variacion1.score(X_test_processed, y_test)
print("Accuracy con max_leaf_nodes=2:", accuracy_variacion1)

# Variación 2: max_leaf_nodes = 10
classifier_variacion2 = DecisionTreeClassifier(
    criterion=mejor_hiperparametros["criterion"],
    splitter=mejor_hiperparametros["splitter"],
    max_depth=mejor_hiperparametros["max_depth"],
    random_state=mejor_hiperparametros["random_state"],
    max_leaf_nodes=10,
)

classifier_variacion2.fit(X_train_processed, y_train)
accuracy_variacion2 = classifier_variacion2.score(X_test_processed, y_test)
print("Accuracy con max_leaf_nodes=10:", accuracy_variacion2)




Tabla de precisión de árboles punto 4
   Profundidad máxima  Precisión
0                   1   0.929825
1                   2   0.929825
2                   3   0.956140
3                   4   0.947368
4                   5   0.964912
5                   6   0.964912
6                   7   0.956140
7                   8   0.964912
8                   9   0.964912
9                  10   0.964912 

Tabla de precisión de árboles punto 6
   Profundidad máxima  Precisión
0                   1   0.929825
1                   2   0.929825
2                   3   0.973684
3                   4   0.964912
4                   5   0.964912
5                   6   0.964912
6                   7   0.991228
7                   8   0.991228
8                   9   0.991228
9                  10   0.991228 

Tabla de precisión de árboles punto 8
   Profundidad máxima  Precisión
0                   1   0.973684
1                   2   0.982456
2                   3   0.973684
3                   4   

# <center>💟CONCLUSION💟</center>

Podemos concluir que cambiando el max_leaf_nodes a los valores de 2 y 10 se llega a una Accuracy de 0.9649122807017544 y 0.9736842105263158 respectivamente, lo cual nos da un resultado mas insatisfactible que nuestro mejor arbol que posee un Accuracy de 0.9912280701754386